examples/platform/silabs/SiWx917/SiWxWifiInterface.cpp

```diff
@@ -69,89 +77,368 @@ extern "C" {
+
+    // Initiate and program the key required for TRNG hardware engine
+    status = sl_si91x_trng_program_key((uint32_t *) trngKey, TRNG_KEY_SIZE);
+    VerifyOrReturnError(status == SL_STATUS_OK, status,
+                        ChipLogError(DeviceLayer, "sl_si91x_trng_program_key failed: 0x%lx", static_cast<uint32_t>(status)));
+#endif // SL_MBEDTLS_USE_TINYCRYPT
+
+    wfx_rsi.dev_state |= WFX_RSI_ST_DEV_READY;
+    osSemaphoreRelease(sl_rs_ble_init_sem);
+    return status;
+}
+
+// TODO: this changes will be reverted back after the Silabs WiFi SDK team fix the scan API
+#ifndef EXP_BOARD
+sl_status_t ScanCallback(sl_wifi_event_t event, sl_wifi_scan_result_t * scan_result, uint32_t result_length, void * arg)
+{
+    sl_status_t status = SL_STATUS_OK;
+    if (SL_WIFI_CHECK_IF_EVENT_FAILED(event))
+    {
+        ChipLogError(DeviceLayer, "Scan Netwrok Failed: 0x%lx", *reinterpret_cast<sl_status_t *>(status));
+#if WIFI_ENABLE_SECURITY_WPA3_TRANSITION
+        security = SL_WIFI_WPA3;
+#else
+        security = SL_WIFI_WPA_WPA2_MIXED;
+#endif /* WIFI_ENABLE_SECURITY_WPA3_TRANSITION */
 
-    // Check if timer started
-    if (!osTimerIsRunning(sDHCPTimer))
+        status = SL_STATUS_FAIL;
+    }
+    else
     {
-        ChipLogDetail(DeviceLayer, "CancelDHCPTimer: timer not running");
-        return;
+        security        = static_cast<sl_wifi_security_t>(scan_result->scan_info[0].security_mode);
+        wfx_rsi.ap_chan = scan_result->scan_info[0].rf_channel;
+        memcpy(&wfx_rsi.ap_mac.octet, scan_result->scan_info[0].bssid, BSSID_LEN);
     }
 
-    status = osTimerStop(sDHCPTimer);
-    if (status != osOK)
+    osSemaphoreRelease(sScanCompleteSemaphore);
+    return status;
+}
+#endif
+
+sl_status_t InitiateScan()
+{
+    sl_status_t status = SL_STATUS_OK;
+
+// TODO: this changes will be reverted back after the Silabs WiFi SDK team fix the scan API
+#ifndef EXP_BOARD
+    sl_wifi_ssid_t ssid = { 0 };
+
+    // TODO: this changes will be reverted back after the Silabs WiFi SDK team fix the scan API
+    sl_wifi_scan_configuration_t wifi_scan_configuration = default_wifi_scan_configuration;
+
+    ssid.length = wfx_rsi.sec.ssid_length;
+
+    // TODO: workaround because the string management with the null termination is flawed
+    chip::Platform::CopyString((char *) &ssid.value[0], ssid.length + 1, wfx_rsi.sec.ssid);
+    sl_wifi_set_scan_callback(ScanCallback, NULL);
+
+    osSemaphoreAcquire(sScanInProgressSemaphore, osWaitForever);
+
+    // This is an odd success code?
+    status = sl_wifi_start_scan(SL_WIFI_CLIENT_2_4GHZ_INTERFACE, &ssid, &wifi_scan_configuration);
+    if (status == SL_STATUS_IN_PROGRESS)
     {
-        ChipLogError(DeviceLayer, "CancelDHCPTimer: failed to stop timer: %d", status);
+        osSemaphoreAcquire(sScanCompleteSemaphore, kWifiScanTimeoutTicks);
+        status = SL_STATUS_OK;
     }
+    else
+    {
+        osSemaphoreRelease(sScanInProgressSemaphore);
```

Comment:
Where is `sScanInProgressSemaphore` released in the case of `SL_STATUS_IN_PROGRESS == status`

Reply:
fixed